src/common/metrics/src/monitor/in_mem.rs

```diff
@@ -16,74 +16,105 @@
 //! It is intentionally decoupled from Prometheus.
 
 use std::collections::HashMap;
-use std::sync::Arc;
 use std::sync::atomic::AtomicU64;
+use std::sync::{Arc, Weak};
 
-use parking_lot::RwLock;
-
-use crate::MetricLevel;
+use parking_lot::Mutex;
 
 pub type Count = Arc<AtomicU64>;
 
+pub struct GuardedCount {
+    id: u64,
+    pub count: Count,
+    parent: Weak<Mutex<InnerCountMap>>,
+}
+
+impl GuardedCount {
+    pub fn new(id: u64, parent: &Arc<Mutex<InnerCountMap>>) -> (Count, Self) {
+        let guard = GuardedCount {
+            id,
+            count: Arc::new(AtomicU64::new(0)),
+            parent: Arc::downgrade(parent),
+        };
+        (guard.count.clone(), guard)
+    }
+}
+
+impl Drop for GuardedCount {
+    fn drop(&mut self) {
+        if let Some(parent) = self.parent.upgrade() {
+            let mut map = parent.lock();
+            map.inner.remove(&self.id);
+        } else {
+            tracing::warn!("parent dropped unexpectedly");
+        }
```

Comment:
I think there's no need to warn about this, as there's no resource leakage in any case.

Reply:
Fixed